lib/backend/k8s/resources/node.go

```diff
@@ -282,6 +300,113 @@ func mergeCalicoNodeIntoK8sNode(calicoNode *apiv3.Node, k8sNode *kapiv1.Node) (*
+// labels.
+func mergeCalicoAndK8sLabels(calicoNode *apiv3.Node, k8sNode *kapiv1.Node) error {
+	// Now, copy the Kubernetes Node labels over.  Note: this may overwrite Calico labels of the same name, but that's
+	// consistent with the kube-controllers behavior.
+	var shadowedLabels map[string]string
+	for k, v := range k8sNode.Labels {
+		if calicoNode.Labels == nil {
+			calicoNode.Labels = map[string]string{}
+		}
+		if calValue, ok := calicoNode.Labels[k]; ok {
+			if shadowedLabels == nil {
+				shadowedLabels = map[string]string{}
+			}
+			shadowedLabels[k] = calValue
+		}
+		calicoNode.Labels[k] = v
+	}
+
+	// For consistency with kube-controllers, and so we can correctly round-trip labels, we stash the kubernetes labels
+	// in an annotation.
+	if calicoNode.Annotations == nil {
+		calicoNode.Annotations = map[string]string{}
+	}
+	bytes, err := json.Marshal(k8sNode.Labels)
+	if err != nil {
+		log.WithError(err).Errorf("Error marshalling node labels")
+		return err
+	}
+	calicoNode.Annotations[nodeK8sLabelAnnotation] = string(bytes)
+
+	// To be able to round trip Calico labels, we stash any labels that we've shadowed in an annotation.
+	if shadowedLabels != nil {
+		bytes, err := json.Marshal(shadowedLabels)
+		if err != nil {
+			log.WithError(err).Errorf("Error marshalling node labels")
+			return err
+		}
+		calicoNode.Annotations[nodeShadowedLabelAnnotation] = string(bytes)
+	}
+	return nil
+}
+
+// restoreCalicoLabels tries to undo the transformation done by mergeCalicoLabels.  It returns a copy of the node with
+// Kubernetes labels removed from the Labels field and restores any shadowed Calico labels.  It also filters the
+// bookkeeping annotations out of the copy.
+func restoreCalicoLabels(calicoNode *apiv3.Node) (*apiv3.Node, error) {
+	calicoNode = calicoNode.DeepCopy()
+
+	if rawLabels := calicoNode.Annotations[nodeK8sLabelAnnotation]; rawLabels != "" {
+		// We stashed the k8s labels in an annotation, extract them so we can compare with the combined labels.
+		k8sLabels := map[string]string{}
+		if err := json.Unmarshal([]byte(rawLabels), &k8sLabels); err != nil {
+			log.WithError(err).Error("Failed to unmarshal k8s node labels from " +
+				nodeK8sLabelAnnotation + " annotation")
+			return nil, err
+		}
+		// We stashed any Calico labels that got shadowed (when we calculated the inherited labels) in an annotation,
+		// extract them.
+		shadowedLabels := map[string]string{}
+		rawShadowedLabels := calicoNode.Annotations[nodeShadowedLabelAnnotation]
+		if rawShadowedLabels != "" {
+			if err := json.Unmarshal([]byte(rawShadowedLabels), &shadowedLabels); err != nil {
+				log.WithError(err).Error("Failed to unmarshal Calico labels from " +
+					nodeShadowedLabelAnnotation + " annotation")
+				return nil, err
+			}
+		}
+
+		// Now remove any labels that match the k8s ones.
+		if log.GetLevel() >= log.DebugLevel {
```

Comment:
is this if statement necessary?

Reply:
No, if you see that pattern, it's a performance optimisation.  WithFields can be expensive if it's done in a loop (it has to copy the log entry and create a new fields dict).  Probably borderline here but this function could be used in a migration or a list operation for 10k nodes with 10 labels each so I thought it might be worthwhile (and it's trivial to do).